### PR TITLE
mysql/json:access slice directly instead of using read trait

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
    This PR updates the decode function for `json`.  It accesses slice directly instead of using `read` trait since `read` trait is not sufficient enough.
@BusyJay @breeswish @hicqu PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
please add a description.

Btw, how about the benchmark?

**Comment by AndreMouche** [comment]
The benchmark result is in our internal jira since the source code is not convenient to public here. @siddontang

**Comment by AndreMouche** [comment]
benchmark result
```
test json::bench::bench_json_decode_binary      ... bench:  23,204,380 ns/iter (+/- 2,596,274)
test json::bench::bench_json_decode_binary_read ... bench:  24,761,830 ns/iter (+/- 2,947,690)
test json::binary::bench_json_decode            ... bench:       2,260 ns/iter (+/- 943)
test json::binary::bench_json_decode_read       ... bench:       2,606 ns/iter (+/- 721)
```

source code:

```
#[bench]
fn bench_json_decode_binary(b: &mut Bencher) {
    // data from https://download.pingcap.org/resources/world_bank.json.tar.gz
    let binaries = load_test_jsons()
        .unwrap()
        .into_iter()
        .map(|t| t.parse::<Json>().unwrap())
        .map(|j| {
            let mut buf = Vec::new();
            buf.encode_json(&j).unwrap();
            buf
        })
        .collect::<Vec<Vec<u8>>>();
    b.iter(|| {
        for binary in &binaries {
            Json::decode(&mut binary.as_slice()).unwrap();
        }
    });
}

#[bench]
fn bench_json_decode_binary_read(b: &mut Bencher) {
    //data from https://download.pingcap.org/resources/world_bank.json.tar.gz
    let binaries = load_test_jsons()
        .unwrap()
        .into_iter()
        .map(|t| t.parse::<Json>().unwrap())
        .map(|j| {
            let mut buf = Vec::new();
            buf.encode_json(&j).unwrap();
            buf
        })
        .collect::<Vec<Vec<u8>>>();
    b.iter(|| {
        for binary in &binaries {
            binary.as_slice().decode_json().unwrap();
        }
    });
}

#[bench]
fn bench_json_decode_read(b: &mut Bencher) {
        let jstr1 =
            r#"{"aaaaaaaaaaa": [1, "2", {"aa": "bb"}, 4.0], "bbbbbbbbbb": true, "ccccccccc": "d"}"#;
        let j1: Json = jstr1.parse().unwrap();
        let jstr2 = r#"[{"a": 1, "b": true}, 3, 3.5, "hello, world", null, true]"#;
        let j2: Json = jstr2.parse().unwrap();

        let json_nil = Json::None;
        let json_bool = Json::Boolean(true);
        let json_int = Json::I64(30);
        let json_uint = Json::U64(30);
        let json_double = Json::Double(3.24);
        let json_str = Json::String(String::from("hello, 世界"));
        let test_cases = vec![
            json_nil,
            json_bool,
            json_int,
            json_uint,
            json_double,
            json_str,
            j1,
            j2,
        ];
        let mut buf = vec![];
        let mut res = vec![Json::None;test_cases.len()];
        test_cases.iter().for_each(|v|buf.encode_json(v).unwrap());
        b.iter(|| {
         let mut slice = buf.as_slice();
          res[0] = slice.decode_json().unwrap();
          res[1] = slice.decode_json().unwrap();
          res[2] = slice.decode_json().unwrap();
          res[3] = slice.decode_json().unwrap();
          res[4] = slice.decode_json().unwrap();
          res[5] = slice.decode_json().unwrap();
          res[6] = slice.decode_json().unwrap();
          res[7] = slice.decode_json().unwrap();
        });
        for id in 0..test_cases.len() {
            let input_str = test_cases[id].to_string();
            let output_str = res[id].to_string();
            assert_eq!(input_str, output_str);
        }
    }

#[bench]
fn bench_json_decode(b: &mut Bencher) {
        let jstr1 =
            r#"{"aaaaaaaaaaa": [1, "2", {"aa": "bb"}, 4.0], "bbbbbbbbbb": true, "ccccccccc": "d"}"#;
        let j1: Json = jstr1.parse().unwrap();
        let jstr2 = r#"[{"a": 1, "b": true}, 3, 3.5, "hello, world", null, true]"#;
        let j2: Json = jstr2.parse().unwrap();

        let json_nil = Json::None;
        let json_bool = Json::Boolean(true);
        let json_int = Json::I64(30);
        let json_uint = Json::U64(30);
        let json_double = Json::Double(3.24);
        let json_str = Json::String(String::from("hello, 世界"));
        let test_cases = vec![
            json_nil,
            json_bool,
            json_int,
            json_uint,
            json_double,
            json_str,
            j1,
            j2,
        ];
        let mut buf = vec![];
        let mut res = vec![Json::None;test_cases.len()];
        test_cases.iter().for_each(|v|buf.encode_json(v).unwrap());
        b.iter(|| {
         let mut slice = buf.as_slice();
          res[0] = Json::decode(&mut slice).unwrap();
          res[1] = Json::decode(&mut slice).unwrap();
          res[2] = Json::decode(&mut slice).unwrap();
          res[3] = Json::decode(&mut slice).unwrap();
          res[4] = Json::decode(&mut slice).unwrap();
          res[5] = Json::decode(&mut slice).unwrap();
          res[6] = Json::decode(&mut slice).unwrap();
          res[7] = Json::decode(&mut slice).unwrap();
        });
        for id in 0..test_cases.len() {
            let input_str = test_cases[id].to_string();
            let output_str = res[id].to_string();
            assert_eq!(input_str, output_str);
        }
}

```

**Comment by hicqu** [comment]
LGTM.

**Comment by AndreMouche** [comment]
friendly ping @breeswish @BusyJay

**Comment by AndreMouche** [comment]
friendly ping @breeswish @BusyJay

File: src/coprocessor/codec/mysql/json/binary.rs [hunk 1]
```diff
@@ -223,126 +222,125 @@ pub trait JsonEncoder: NumberEncoder {
...
             let key_real_offset = number::decode_u32_le(&mut key_entries_data)?;
             let key_len = number::decode_u16_le(&mut key_entries_data)?;
-            let key_data = &data[key_offset..(key_offset + key_len as usize)];
+            let key_data = read_slice(&mut data, key_len as usize)?;
             let key = String::from(str::from_utf8(key_data).unwrap());
-            let value = value_entries_data.decode_json_item(&data[key_offset..], key_real_offset)?;
+            let value = Json::decode_item(
+                &mut value_entries_data,
+                data,
+                key_real_offset + u32::from(key_len),
```
Comment (breeswish) [inline src/coprocessor/codec/mysql/json/binary.rs]: why do we need `+ u32::from(key_len)`?

Comment (AndreMouche) [reply src/coprocessor/codec/mysql/json/binary.rs]: Since L265 has already moved the data forward `key_len`

Comment (breeswish) [reply src/coprocessor/codec/mysql/json/binary.rs]: oh. I got it.

**Comment by breeswish** [comment]
LGTM

**Pull request merged**